File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int median = 0;
        int middle = (nums1.length + nums2.length + 1) >> 1;
        int i = 0, j = 0;
    
        for (int k = 0; k < middle; k++) {
            if (nums1.length == i) {
                median = nums2[j++];
            } else if (nums2.length == j) {
                median = nums1[i++];
            } else if (nums1[i] > nums2[j]) {
                median = nums2[j++];
            } else {
                median = nums1[i++];
            }
        }

        System.out.println(middle + ", " + i + ", " + j + ", " + median);

        if (middle < (nums1.length + nums2.length + 2) >>1) {
            int peer;
            if (nums1.length == i) {
                peer = nums2[j];
            } else if (nums2.length == j) {
                peer = nums1[i];
            } else if (nums1[i] > nums2[j]) {
                peer = nums2[j];
            } else {
                peer = nums1[i];
            }

            System.out.println(((nums1.length + nums2.length + 2) >> 1) + ", " + peer);

            return (double)(median + peer) / 2;
        } else {
            return median;
        }
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
    int getSmaller(vector<int>& nums1, int& idx1, vector<int>& nums2, int& idx2) {
        if (nums1.size() == idx1) {
            return nums2[idx2++];
        }

        if (nums2.size() == idx2) {
            return nums1[idx1++];
        }

        return nums1[idx1] > nums2[idx2]? nums2[idx2++]: nums1[idx1++];
    }

public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int count, idx1, idx2, val;
        for (idx1 = 0, idx2 = 0, count = 0;
            count < (nums1.size() + nums2.size() + 1) / 2; count++) {
            val = getSmaller(nums1, idx1, nums2, idx2);
        }

        return nums1.size() + nums2.size() == count * 2? 
            double(val + getSmaller(nums1, idx1, nums2, idx2)) / 2: val;
    }
};